File: android-offline/src/com/apolos/shieldlite/DnsFilter.java
package com.apolos.shieldlite;

import android.content.Context;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashSet;
import java.util.Set;

/** Domain blocklist for the DNS-filtering tunnel: seed list + bundled assets + user file. */
final class DnsFilter {

    private static final String[] SEED = {
            "doubleclick.net", "googleadservices.com", "google-analytics.com",
            "googlesyndication.com", "adservice.google.com", "graph.facebook.com",
            "app-measurement.com", "firebase-settings.crashlytics.com", "ads.yahoo.com",
            "adnxs.com", "scorecardresearch.com", "mopub.com", "unityads.unity3d.com",
            "applovin.com", "chartboost.com", "flurry.com",
    };

    private final Set<String> blocked = new HashSet<>();

    DnsFilter(Context ctx) {
        for (String s : SEED) blocked.add(s);
        try (BufferedReader r = new BufferedReader(
                new InputStreamReader(ctx.getAssets().open("blocklist.txt")))) {
            loadFrom(r);
        } catch (IOException ignored) {
            // no bundled asset — seed list still applies
        }
        File userFile = new File(ctx.getFilesDir(), "user_blocklist.txt");
        if (userFile.exists()) {
            try (BufferedReader r = new BufferedReader(new FileReader(userFile))) {
                loadFrom(r);
            } catch (IOException ignored) { }
        }
    }

    private void loadFrom(BufferedReader r) throws IOException {
        String line;
        while ((line = r.readLine()) != null) {
            int hash = line.indexOf('#');
            if (hash >= 0) line = line.substring(0, hash);
            line = line.trim().toLowerCase();
            if (!line.isEmpty()) {
                if (line.startsWith("*.")) line = line.substring(2);
                blocked.add(line);
            }
        }
    }

    boolean isBlocked(String host) {
        if (host == null || host.isEmpty()) return false;
        String h = host;
        while (true) {
            if (blocked.contains(h)) return true;
            int dot = h.indexOf('.');
            if (dot < 0) return false;
            h = h.substring(dot + 1);
        }
    }

    int size() { return blocked.size(); }
}

File: android/app/src/main/java/com/apolos/shield/vpn/DnsFilter.kt
package com.apolos.shield.vpn

import android.content.Context
import java.io.BufferedReader

/**
 * Domain blocklist for the DNS-filtering tunnel. Ships with a compact seed list
 * of common tracker/ad/malware domains and merges any user rules stored in the
 * app's files dir (one domain per line, "#" comments allowed). Matching is
 * suffix-based so "ads.example.com" is blocked by an "example.com" rule only
 * when that rule is intended as a zone; exact + parent-domain matching is used.
 */
class DnsFilter(context: Context) {

    private val blocked = HashSet<String>()

    init {
        SEED.forEach { blocked.add(it) }
        runCatching {
            context.assets.open("blocklist.txt").bufferedReader().use(::loadFrom)
        }
        val userFile = context.filesDir.resolve("user_blocklist.txt")
        if (userFile.exists()) userFile.bufferedReader().use(::loadFrom)
    }

    private fun loadFrom(reader: BufferedReader) {
        reader.lineSequence().forEach { raw ->
            val line = raw.substringBefore('#').trim().lowercase()
            if (line.isNotEmpty()) blocked.add(line.removePrefix("*."))
        }
    }

    fun isBlocked(host: String?): Boolean {
        if (host.isNullOrEmpty()) return false
        var h = host
        // Walk up the labels: a.b.c.com -> b.c.com -> c.com -> com
        while (true) {
            if (blocked.contains(h)) return true
            val dot = h!!.indexOf('.')
            if (dot < 0) return false
            h = h.substring(dot + 1)
        }
    }

    fun size(): Int = blocked.size

    companion object {
        // Small, well-known seed set; the real weight comes from assets/blocklist.txt.
        private val SEED = setOf(
            "doubleclick.net",
            "googleadservices.com",
            "google-analytics.com",
            "googlesyndication.com",
            "adservice.google.com",
            "graph.facebook.com",
            "app-measurement.com",
            "firebase-settings.crashlytics.com",
            "ads.yahoo.com",
            "adnxs.com",
            "scorecardresearch.com",
            "mopub.com",
            "unityads.unity3d.com",
            "applovin.com",
            "chartboost.com",
            "flurry.com",
        )
    }
}
